include/envoy/network/listen_socket.h

```diff
@@ -34,5 +34,65 @@ class ListenSocket {
 typedef std::unique_ptr<ListenSocket> ListenSocketPtr;
 typedef std::shared_ptr<ListenSocket> ListenSocketSharedPtr;
 
+/**
+ * An abstract accepted socket.
+ */
+class AcceptedSocket {
+public:
+  virtual ~AcceptedSocket() {}
+
+  /**
+   * @return the address that the socket was received at, or an original destination address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr localAddress() const PURE;
+
+  /**
+   * Reset the original destination address of the socket to a different address than the one
+   * the socket was accepted at.
+   */
+  virtual void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) PURE;
+
+  /**
+   * @return true if the local address has been reset.
+   */
+  virtual bool localAddressReset() const PURE;
+
+  /**
+   * @return the address that the socket was received from, or an original source address if
+   * applicable.
+   */
+  virtual Address::InstanceConstSharedPtr remoteAddress() const PURE;
+
+  /**
+   * Set the original source address of the socket
+   */
+  virtual void resetRemoteAddress(const Address::InstanceConstSharedPtr& remote_address) PURE;
+
+  /**
+   * @return fd the accepted socket's file descriptor.
+   */
+  virtual int fd() const PURE;
+
+  /**
+   * Transfer ownership of the file descriptor to the caller, so that the underlying socket will
+   * not be closed on delete.
+   */
+  virtual int takeFd() PURE;
+
+  /**
+   * Clear 'reset' state so that the socket can be used again with a new listener.
+   */
+  virtual void clearReset() PURE;
+
+  /**
+   * Close the underlying socket.
+   */
+  virtual void close() PURE;
```

Comment:
Do we need an explicit `close()` method? Can this be handled in destruction? Even in the case of deferred deletion it shouldn't matter.

Reply:
Right, I'll remove the explicit close().

Reply:
Had to reintroduce this in order to remove takeFd(), which was more offensive.